src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
Here's one possibility, but i'm not sure i like how htis looks:
(note: the names are obviously not final)

```c#
abstract class CodeRefactoringProvider2 : CodeRefactoringProvider
{
    public abstract Task<object> GetUIStateAsync(Document document, TextSpan span, CancellationToken ct);
    public abstract Task ComputeRefactoringsAsync(CodeRefactoringContext context, object state);
}
```

Is that what you guys were thinking?